genesis/engine/coupler.py

```diff
@@ -1,9 +1,14 @@
+from typing import TYPE_CHECKING
```

Comment:
System / built-in dependencies are usually separated from the others by a blank line. But no big deal.

Reply:
Yeah, we should consider using `isort` or `ruff` to handle this for us I feel.

Reply:
Yes, let's setup this soon. `ruff` would be nice.

Reply:
Awesome! 🙌